### DISCUSSION TITLE
Better Shadows

### DISCUSSION THREAD

**Discussion by CatCode79** [Davidster/ikari#6] [Ideas]
I open a discussion on how to improve shadows by collecting some notes from the [Performance Issue PR](https://github.com/Davidster/wgpu-sandbox/issues/3)

The current implementation is based on: [Point Shadows](https://learnopengl.com/Advanced-Lighting/Shadows/Point-Shadows)

- shadows are taking up something like 7ms per frame which is too slow, it's something I want to investigate a bit later but I found some interesting material about doing shadows efficiently here in the "Atlases" section https://godotengine.org/article/godot-3-renderer-design-explained.

- reading a bit about the shadow atlas in Godot 3 I discovered that allows rendering multiple lights in a single pass.
Instead, Volumetric Shadows techniques must calculate every single light in the scene. This is only feasible if there are few vertices and/or lights.
They also allow you to have sharp and precise shadows, but to make soft shadows you have to do other calculations.

Other techniques:
- [Cascaded Shadow Mapping](https://doc.babylonjs.com/features/featuresDeepDive/lights/shadows_csm)
- [Percentage Closer Filtering (PCF)](https://doc.babylonjs.com/features/featuresDeepDive/lights/mathShadows)
- Percentage Closer Soft Shadows (PCSS) ^the same link above

**Comment by Davidster** [comment]
adding this thread I found last week on reddit as an additional resource: https://www.reddit.com/r/GraphicsProgramming/comments/lf7yes/how_to_optimize_shadows_in_your_rendering_engine

**Comment by CatCode79** [comment]
Interesting!
Just as I tested the hdr I noticed that the character's shadow was cast on the wall even from a long distance, there was also a noticeable delay (i.e. while I touched the floor the shadow was still "falling" and it was halfway up the wall). I thought that to avoid these "oddities" you could avoid drawing the shadows beyond a certain distance, just like suggested in reddit's post note #6.

**Reply by Davidster** [reply]
Yup! I should add frustum culling to the shadow mapping pipeline and adjust the frustum so things that are very far don't get drawn, I think this will give a big performance boost!! I'm excited to try implementing it.

**Reply by Davidster** [reply]
I guess the max distance of the shadow map should also depend on the brightness of the light since that affects how far away the intensity of the light falls of to 0

**Reply by CatCode79** [reply]
Yes! frustum culling is probably one of the best techniques to increase performance (together with LOD and occlusion culling). There are so many interesting tricks, I remember an article on Zelda Breath of the Wild, where she explained that the distant trees were actually textures and not 3d models

**Reply by CatCode79** [reply]
I am always amazed by long shadows in sunsets, apparently the real world has no computational problems :D
In videogames I don't think there are often, except for "mega-structures" (buildings or mountains). Perhaps only long shadows are calculated for static elements in the scene, while moving models can do without them as, being smaller, their long shadow would typically be dimmed by ambient light or other lights, thus also solving a problem computational.

**Reply by Davidster** [reply]
yeah, for large worlds I'd be curious to know how they do it

**Comment by CatCode79** [comment]
I recently discovered this crate: [xatlas-rs](https://github.com/Jasper-Bekkers/xatlas-rs)
Mesh charting, parameterization and atlas packing. Suitable for generating unique texture coordinates for baking lightmaps.

**Reply by Davidster** [reply]
Seems cool, is it supposed to be used at runtime or a tool for processing the assets offline? I would assume that blender has something similar built into it. It almost seems more like a tool for artists or as a plugin for a modelling software.

**Reply by CatCode79** [reply]
you're right, it looks more like a blender-like-tool than something to use as a real time library. I was hoping it could be used as a semi-finished product for shadow atlases, but I still have a vague idea of how these techniques can be implemented to say for sure
